File: java/smr-collections-common/src/test/java/net/knego/hiperf/smrcollections/SnapshotTest.java
package net.knego.hiperf.smrcollections;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.Arrays;
import net.knego.hiperf.common.SmrConfig;
import org.junit.jupiter.api.Test;

class SnapshotTest {
    private static SmrConfig cfg() {
        return new SmrConfig(4096, 64, 1, 0, 2000, 0, 0, 4096, 200000, 20000, 100);
    }

    private static Book build(SmrConfig c, int n) {
        Book b = new Book(c);
        Workload.SplitMix rng = new Workload.SplitMix(Workload.SEED);
        Workload.Insert ins = new Workload.Insert();
        for (int i = 0; i < n; i++) {
            Workload.nextInsert(rng, i, c.levels(), c.tick(), c.priceMin(), ins);
            b.insert(ins.orderId, ins.price, ins.qty, ins.side);
        }
        return b;
    }

    @Test
    void roundTripPreservesQueries() {
        SmrConfig c = cfg();
        Book b = build(c, c.steady());
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int len = s.encode(b);
        byte[] img = Arrays.copyOf(s.backing(), len);
        Book r = Snapshotter.restore(img, len, c);
        assertEquals(b.bestBid(), r.bestBid());
        assertEquals(b.bestAsk(), r.bestAsk());
        assertEquals(b.hwm(), r.hwm());
        for (long id = 1; id <= c.steady(); id++) {
            assertEquals(b.getSlot(id), r.getSlot(id));
        }
        for (int t = 0; t < c.levels(); t++) {
            assertEquals(b.levelQty((byte) 0, t), r.levelQty((byte) 0, t));
            assertEquals(b.levelQty((byte) 1, t), r.levelQty((byte) 1, t));
        }
    }

    @Test
    void deterministicBytes() {
        SmrConfig c = cfg();
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int l1 = s.encode(build(c, c.steady()));
        byte[] a = Arrays.copyOf(s.backing(), l1);
        int l2 = s.encode(build(c, c.steady()));
        byte[] bb = Arrays.copyOf(s.backing(), l2);
        assertArrayEquals(a, bb);
    }

    private static Book buildBookWithCancels(SmrConfig c, int n, int cancelEvery) {
        Book b = new Book(c);
        Workload.SplitMix rng = new Workload.SplitMix(Workload.SEED);
        Workload.Insert ins = new Workload.Insert();
        for (int i = 0; i < n; i++) {
            Workload.nextInsert(rng, i, c.levels(), c.tick(), c.priceMin(), ins);
            b.insert(ins.orderId, ins.price, ins.qty, ins.side);
            if (i % cancelEvery == cancelEvery - 1) {
                b.cancel(ins.orderId);
            }
        }
        return b;
    }

    private static SmrConfig snapCfg() {
        return new SmrConfig(4096, 64, 1L, 0L, 2000, 0, 0, 4096, 200_000, 20_000, 100);
    }

    @Test
    void roundTripPreservesFreeListOrder() {
        SmrConfig c = snapCfg();
        Book b = buildBookWithCancels(c, c.steady(), 4);
        assertNotEquals(Book.NIL, b.freeHead, "test needs a non-empty free list");
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int n = s.encode(b);
        byte[] img = java.util.Arrays.copyOf(s.backing(), n);
        Book r = Snapshotter.restore(img, n, c);
        assertEquals(walkFree(b), walkFree(r), "free list order survives exactly");
    }

    private static java.util.List<Integer> walkFree(Book b) {
        java.util.List<Integer> out = new java.util.ArrayList<>();
        for (int slot = b.freeHead; slot != Book.NIL; slot = b.pool[slot].next) {
            out.add(slot);
        }
        return out;
    }

    @Test
    void restoreAfterCancelsReencodesIdentically() {
        SmrConfig c = snapCfg();
        Book b = buildBookWithCancels(c, c.steady(), 4);
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int n1 = s.encode(b);
        byte[] first = java.util.Arrays.copyOf(s.backing(), n1);
        Book r = Snapshotter.restore(first, n1, c);
        Snapshotter s2 = new Snapshotter(4 * 1024 * 1024);
        int n2 = s2.encode(r);
        byte[] second = java.util.Arrays.copyOf(s2.backing(), n2);
        assertArrayEquals(first, second);
    }

    @Test
    void freedSlotsStayOutOfTheIdMap() {
        SmrConfig c = snapCfg();
        Book b = buildBookWithCancels(c, c.steady(), 4);
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int n = s.encode(b);
        Book r = Snapshotter.restore(java.util.Arrays.copyOf(s.backing(), n), n, c);
        for (int slot = 0; slot < b.hwm(); slot++) {
            long id = b.pool[slot].orderId;
            if (id != 0) {
                assertEquals(slot, r.getSlot(id), "live order " + id + " keeps its slot");
            } else {
                assertEquals(0, r.pool[slot].orderId, "slot " + slot + " stays marked free");
            }
        }
        assertNull(r.ids.get(0L), "orderId 0 must never be a key");
    }

    @Test
    void restoreRejectsCapacityMismatch() {
        SmrConfig c = snapCfg();
        Book b = buildBookWithCancels(c, c.steady(), 4);
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int n = s.encode(b);
        byte[] img = java.util.Arrays.copyOf(s.backing(), n);
        SmrConfig smaller = new SmrConfig(2048, 64, 1L, 0L, 2000, 0, 0, 2048, 200_000, 20_000, 100);
        assertThrows(IllegalArgumentException.class, () -> Snapshotter.restore(img, n, smaller));
    }
}

File: go/internal/smrcoll/snapshot_test.go
package smrcoll

import (
	"bytes"
	"testing"

	"github.com/peterknego/hi-perf-cmp/go/internal/bench"
)

func buildBook(c bench.SmrConfig, n int) *Book {
	b := NewBook(c)
	rng := NewSplitMix(SmrSeed)
	for i := 0; i < n; i++ {
		ins := NextInsert(rng, i, c.Levels, c.Tick, c.PriceMin)
		b.Insert(ins.OrderID, ins.Price, ins.Qty, ins.Side)
	}
	return b
}

func snapCfg() bench.SmrConfig {
	return bench.SmrConfig{Cap: 4096, Levels: 64, Tick: 1, PriceMin: 0, Steady: 2000, Warmup: 0, Iters: 0}
}

func TestSnapshotRoundTrip(t *testing.T) {
	c := snapCfg()
	b := buildBook(c, c.Steady)
	s := NewSnapshotter()
	img := append([]byte(nil), s.Encode(b)...) // copy: buffer is reused
	r, err := Restore(img, c)
	if err != nil {
		t.Fatalf("restore: %v", err)
	}
	if r.BestBidTick() != b.BestBidTick() || r.BestAskTick() != b.BestAskTick() || r.HwmVal() != b.HwmVal() {
		t.Fatalf("header mismatch after restore")
	}
	for id := int64(1); id <= int64(c.Steady); id++ {
		if r.GetSlot(id) != b.GetSlot(id) {
			t.Fatalf("slot mismatch for id %d", id)
		}
	}
	for tk := uint32(0); tk < c.Levels; tk++ {
		if r.LevelQty(0, tk) != b.LevelQty(0, tk) || r.LevelQty(1, tk) != b.LevelQty(1, tk) {
			t.Fatalf("level qty mismatch at tick %d", tk)
		}
	}
}

func TestSnapshotDeterministic(t *testing.T) {
	c := snapCfg()
	s := NewSnapshotter()
	a := append([]byte(nil), s.Encode(buildBook(c, c.Steady))...)
	b := append([]byte(nil), s.Encode(buildBook(c, c.Steady))...)
	if !bytes.Equal(a, b) {
		t.Fatalf("same ops => bytes must be identical")
	}
}

func buildBookWithCancels(c bench.SmrConfig, n, cancelEvery int) *Book {
	b := NewBook(c)
	rng := NewSplitMix(SmrSeed)
	for i := 0; i < n; i++ {
		ins := NextInsert(rng, i, c.Levels, c.Tick, c.PriceMin)
		b.Insert(ins.OrderID, ins.Price, ins.Qty, ins.Side)
		if i%cancelEvery == cancelEvery-1 {
			b.Cancel(ins.OrderID)
		}
	}
	return b
}

func TestRoundTripPreservesFreeListOrder(t *testing.T) {
	c := bench.SmrConfig{Cap: 4096, Levels: 64, Tick: 1, PriceMin: 0, Steady: 2000, Chunk: 4096, OtrBps: 100}
	b := buildBookWithCancels(c, c.Steady, 4)
	if b.FreeHead == NIL {
		t.Fatal("test needs a non-empty free list")
	}
	s := NewSnapshotter()
	img := append([]byte(nil), s.Encode(b)...)
	r, err := Restore(img, c)
	if err != nil {
		t.Fatalf("restore: %v", err)
	}
	walk := func(bk *Book) []uint32 {
		var out []uint32
		for slot := bk.FreeHead; slot != NIL; slot = bk.Pool[slot].Next {
			out = append(out, slot)
		}
		return out
	}
	got, want := walk(r), walk(b)
	if len(got) != len(want) {
		t.Fatalf("free list length %d, want %d", len(got), len(want))
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("free list[%d] = %d, want %d", i, got[i], want[i])
		}
	}
}

func TestRestoreAfterCancelsReencodesIdentically(t *testing.T) {
	c := bench.SmrConfig{Cap: 4096, Levels: 64, Tick: 1, PriceMin: 0, Steady: 2000, Chunk: 4096, OtrBps: 100}
	b := buildBookWithCancels(c, c.Steady, 4)
	s := NewSnapshotter()
	first := append([]byte(nil), s.Encode(b)...)
	r, err := Restore(first, c)
	if err != nil {
		t.Fatalf("restore: %v", err)
	}
	second := NewSnapshotter().Encode(r)
	if !bytes.Equal(first, second) {
		t.Fatalf("re-encode differs: %d vs %d bytes", len(first), len(second))
	}
}

func TestFreedSlotsStayOutOfTheIDMap(t *testing.T) {
	c := bench.SmrConfig{Cap: 4096, Levels: 64, Tick: 1, PriceMin: 0, Steady: 2000, Chunk: 4096, OtrBps: 100}
	b := buildBookWithCancels(c, c.Steady, 4)
	s := NewSnapshotter()
	r, err := Restore(append([]byte(nil), s.Encode(b)...), c)
	if err != nil {
		t.Fatalf("restore: %v", err)
	}
	for slot := uint32(0); slot < b.Hwm; slot++ {
		id := b.Pool[slot].OrderID
		if id != 0 {
			if got := r.GetSlot(id); got != slot {
				t.Fatalf("live order %d: slot %d, want %d", id, got, slot)
			}
		} else if r.Pool[slot].OrderID != 0 {
			t.Fatalf("slot %d lost its freed marker", slot)
		}
	}
	if got := r.GetSlot(0); got != NIL {
		t.Fatalf("OrderID 0 must never be a key, got slot %d", got)
	}
}

func TestRestoreRejectsCapacityMismatch(t *testing.T) {
	c := bench.SmrConfig{Cap: 4096, Levels: 64, Tick: 1, PriceMin: 0, Steady: 2000, Chunk: 4096, OtrBps: 100}
	b := buildBookWithCancels(c, c.Steady, 4)
	img := NewSnapshotter().Encode(b)
	smaller := c
	smaller.Cap = 2048
	if _, err := Restore(img, smaller); err == nil {
		t.Fatal("restoring into a smaller-capacity build must fail loudly")
	}
}
